ggml.c

```diff
@@ -13966,6 +14059,49 @@ typedef pthread_t ggml_thread_t;
 
 #endif
 
+#ifdef __linux__
+void set_numa_thread_affinity(int thread_n, int n_threads)
+{
+    if (!ggml_is_numa()) { return; }
+    // run thread on node_num thread_n / (threads per node)
+    int node_num = thread_n / (n_threads / ggml_numa.n_nodes);
+    struct ggml_numa_node *node = &ggml_numa.nodes[node_num];
+    size_t setsize = CPU_ALLOC_SIZE(ggml_numa.total_cpus);
+    cpu_set_t *cpus = CPU_ALLOC(ggml_numa.total_cpus);
+    CPU_ZERO_S(setsize, cpus);
+    for (size_t i = 0; i < node->n_cpus; ++i) {
+        CPU_SET_S(node->cpus[i], setsize, cpus);
+    }
+    int rv = pthread_setaffinity_np(pthread_self(), setsize, cpus);
+    if (rv) {
+            fprintf(stderr, "warning: pthread_setaffinity_np() failed: %s\n",
+                    strerror(rv));
+    }
+    CPU_FREE(cpus);
+}
+void clear_numa_thread_affinity(void)
```

Comment:
This function does not seem to be used

Reply:
The workers don't need it because they terminate. In the first attempt I wasn't setting affinity for the main thread, which seems to be fine because then it runs on the last remaining core. When I tried setting it performance was degraded if it stayed set after eval, so I wrote that to clear it when the worker threads were joined. That solved the performance degradation, but setting it and clearing it seemed to perform the same as not setting it to begin with, so then I didn't use it.